includes/services/extended/nextdoor.php

```diff
@@ -0,0 +1,176 @@
+		unset( $params['client_secret'] );
+		$params['redirect_uri']  = $this->redirect_uri;
+		// $params[''] = ($params['code'] . '=');
+		return $params;
+	}
+
+	function verify_token_post_params( $params ) {
+		$params['headers'] = $this->get_basic_auth();
+
+		return $params;
+	}
+
+	function fix_access_token_meta( $access_token, $token ) {
+		error_log('testing fix access token meta');
+		if ( 'nextdoor' !== $access_token->get_name() ) {
+			return $access_token;
+		}
+
+		return new Keyring_Access_Token(
+			$this->get_name(),
+			$token['access_token'],
+			array_merge( $access_token->get_meta(), $this->get_token()->get_meta() ) // refresh_token has been updated, and we want to make sure we store it
+		);
+	}
+
+	function get_basic_auth() {
+		return array( 'Authorization' => 'Basic ' . base64_encode( $this->key . ':' . $this->secret ) );
+	}
+
+	function build_token_meta( $token ) {
+		$decoded_id_token = json_decode(base64_decode(explode('.', $token['id_token'])[1]));
+		error_log('testing build_token_meta id token' . print_r($decoded_id_token->sub, true));
+		$meta = array(
+			'account_id'      => $decoded_id_token->sub,
+			'expires'       => time() + $token['expires_in'],
+		);
+
+		$this->set_token(
+			new Keyring_Access_Token(
+				$this->get_name(),
+				$token['access_token'],
+				$meta
+			)
+		);
+		error_log('testing build_token_meta meta' . print_r($meta, true));
+
+
+		return apply_filters( 'keyring_access_token_meta', $meta, $this->get_name(), $token, $this );
+	}
+
+	function get_display( Keyring_Access_Token $token ) {
+		return $token->get_meta( 'name' );
+	}
+
+	function refresh_token() {
+		Keyring_Util::debug( 'Nextdoor::refresh_token()' );
+		// Request a new token, using the refresh_token
+		$token = $this->get_token();
+		$meta  = $token->get_meta();
+		if ( empty( $meta['refresh_token'] ) ) {
+			return false;
+		}
+
+		// Don't bother if this token is valid for a while
+		if ( ! $token->is_expired( 20 ) ) {
+			return;
+		}
+
+		// Refresh our access token
+		$response = wp_remote_post(
+			$this->refresh_url,
+			array(
+				'method'  => $this->refresh_method,
+				'headers' => $this->get_basic_auth(),
+				'body'    => array(
+					'grant_type'    => 'refresh_token',
+					'refresh_token' => $meta['refresh_token'],
+				),
+			)
+		);
+		Keyring_Util::debug( $response );
+
+		if ( 200 !== wp_remote_retrieve_response_code( $response ) ) {
+			return false;
+		}
+
+		$return                = json_decode( wp_remote_retrieve_body( $response ) );
+		$meta['refresh_token'] = $return->refresh_token;
+		$meta['expires']       = time() + $return->expires_in;
+		$access_token          = new Keyring_Access_Token(
+			$this->get_name(),
+			$return->access_token,
+			$meta,
+			$token->get_uniq_id()
+		);
+
+		// Update store, and switch to new token
+		$this->store->update( $access_token );
+		$this->set_token( $access_token );
+	}
+
+	// Need to potentially refresh token before each request
```

Comment:
There should be a recurring job that automatically refreshes the token (at least once before it expires). No need to refresh before every request.

Reply:
It will only refresh the token if it's about to expire. That's [what's happening above](https://github.com/beaulebens/keyring/pull/113/files#diff-4fd3c9fb59cef295050b2c83cfd2918448016df9c1da6562e629f1309df598fbR121-R124)